In [2]:
import json
import pandas as pd

In [3]:
test_depparse = pd.read_json('Restaurants_Test_Gold_depparsed.json')
test_biaffine = pd.read_json('Restaurants_Test_Gold_biaffine_depparsed_with_energy.json')
test_biaffine_energy = pd.read_json('Restaurants_Test_Gold_biaffine_depparsed_with_energy.json')

In [4]:
test_depparse.head()

sentence  \
0                    The bread is top notch as well.   
1  I have to say they have one of the fastest del...   
2        Food is always fresh and hot- ready to eat!   
3      Did I mention that the coffee is OUTSTANDING?   
4  Certainly not the best sushi in New York, howe...   

                                              tokens  \
0          [The, bread, is, top, notch, as, well, .]   
1  [I, have, to, say, they, have, one, of, the, f...   
2  [Food, is, always, fresh, and, hot, -, ready, ...   
3  [Did, I, mention, that, the, coffee, is, OUTST...   
4  [Certainly, not, the, best, sushi, in, New, Yo...   

                                                tags  \
0                   [DT, NN, VBZ, JJ, NN, RB, RB, .]   
1  [PRP, VBP, TO, VB, PRP, VBP, CD, IN, DT, JJS, ...   
2       [NNP, VBZ, RB, JJ, CC, JJ, :, JJ, TO, VB, .]   
3            [VBD, PRP, VB, IN, DT, NN, VBZ, NNP, .]   
4  [RB, RB, DT, JJS, NN, IN, NNP, NNP, ,, RB, ,, ...   

                                        dependencies  \
0  [[root, 0, 5], [det, 2, 1], [nsubj, 5, 2], [co...   
1  [[root, 0, 2], [nsubj, 2, 1], [nsubj:xsubj, 4,...   
2  [[root, 0, 4], [nsubj, 4, 1], [nsubj, 6, 1], [...   
3  [[root, 0, 3], [aux, 3, 1], [nsubj, 3, 2], [ma...   
4  [[root, 0, 15], [advmod, 5, 1], [neg, 5, 2], [...   

               aspect_sentiment     from_to  
0           [[bread, positive]]    [[1, 2]]  
1  [[delivery times, positive]]  [[10, 12]]  
2            [[Food, positive]]    [[0, 1]]  
3          [[coffee, positive]]    [[5, 6]]  
4           [[place, positive]]  [[18, 19]]

In [5]:
test_biaffine.head()

sentence  \
0                    The bread is top notch as well.   
1  I have to say they have one of the fastest del...   
2        Food is always fresh and hot- ready to eat!   
3      Did I mention that the coffee is OUTSTANDING?   
4  Certainly not the best sushi in New York, howe...   

                                              tokens  \
0          [The, bread, is, top, notch, as, well, .]   
1  [I, have, to, say, they, have, one, of, the, f...   
2  [Food, is, always, fresh, and, hot-, ready, to...   
3  [Did, I, mention, that, the, coffee, is, OUTST...   
4  [Certainly, not, the, best, sushi, in, New, Yo...   

                                                tags  \
0                   [DT, NN, VBZ, JJ, NN, RB, RB, .]   
1  [PRP, VBP, TO, VB, PRP, VBP, CD, IN, DT, JJS, ...   
2           [NN, VBZ, RB, JJ, CC, RB, JJ, TO, VB, .]   
3             [VBD, PRP, VB, IN, DT, NN, VBZ, JJ, .]   
4  [RB, RB, DT, JJS, NN, IN, NNP, NNP, ,, RB, ,, ...   

                              predicted_dependencies  \
0  [det, nsubj, cop, amod, root, advmod, advmod, ...   
1  [nsubj, root, aux, xcomp, nsubj, ccomp, dobj, ...   
2  [nsubj, cop, advmod, root, cc, advmod, conj, a...   
3  [aux, nsubj, root, mark, det, nsubj, cop, ccom...   
4  [dep, dep, det, amod, nsubj, prep, nn, pobj, p...   

                                     predicted_heads  \
0                           [2, 5, 5, 5, 0, 7, 5, 5]   
1  [2, 0, 4, 2, 6, 4, 6, 7, 12, 12, 12, 8, 12, 15...   
2                     [4, 4, 4, 0, 4, 7, 4, 9, 7, 4]   
3                        [3, 3, 0, 8, 6, 8, 8, 3, 3]   
4  [2, 5, 5, 5, 15, 5, 8, 6, 15, 15, 15, 15, 15, ...   

                                        dependencies  \
0  [[det, 2, 1], [nsubj, 5, 2], [cop, 5, 3], [amo...   
1  [[nsubj, 2, 1], [root, 0, 2], [aux, 4, 3], [xc...   
2  [[nsubj, 4, 1], [cop, 4, 2], [advmod, 4, 3], [...   
3  [[aux, 3, 1], [nsubj, 3, 2], [root, 0, 3], [ma...   
4  [[dep, 2, 1], [dep, 5, 2], [det, 5, 3], [amod,...   

               aspect_sentiment     from_to  
0           [[bread, positive]]    [[1, 2]]  
1  [[delivery times, positive]]  [[10, 12]]  
2            [[Food, positive]]    [[0, 1]]  
3          [[coffee, positive]]    [[5, 6]]  
4           [[place, positive]]  [[18, 19]]

In [6]:
test_biaffine_energy.head()

sentence  \
0                    The bread is top notch as well.   
1  I have to say they have one of the fastest del...   
2        Food is always fresh and hot- ready to eat!   
3      Did I mention that the coffee is OUTSTANDING?   
4  Certainly not the best sushi in New York, howe...   

                                              tokens  \
0          [The, bread, is, top, notch, as, well, .]   
1  [I, have, to, say, they, have, one, of, the, f...   
2  [Food, is, always, fresh, and, hot-, ready, to...   
3  [Did, I, mention, that, the, coffee, is, OUTST...   
4  [Certainly, not, the, best, sushi, in, New, Yo...   

                                                tags  \
0                   [DT, NN, VBZ, JJ, NN, RB, RB, .]   
1  [PRP, VBP, TO, VB, PRP, VBP, CD, IN, DT, JJS, ...   
2           [NN, VBZ, RB, JJ, CC, RB, JJ, TO, VB, .]   
3             [VBD, PRP, VB, IN, DT, NN, VBZ, JJ, .]   
4  [RB, RB, DT, JJS, NN, IN, NNP, NNP, ,, RB, ,, ...   

                              predicted_dependencies  \
0  [det, nsubj, cop, amod, root, advmod, advmod, ...   
1  [nsubj, root, aux, xcomp, nsubj, ccomp, dobj, ...   
2  [nsubj, cop, advmod, root, cc, advmod, conj, a...   
3  [aux, nsubj, root, mark, det, nsubj, cop, ccom...   
4  [dep, dep, det, amod, nsubj, prep, nn, pobj, p...   

                                     predicted_heads  \
0                           [2, 5, 5, 5, 0, 7, 5, 5]   
1  [2, 0, 4, 2, 6, 4, 6, 7, 12, 12, 12, 8, 12, 15...   
2                     [4, 4, 4, 0, 4, 7, 4, 9, 7, 4]   
3                        [3, 3, 0, 8, 6, 8, 8, 3, 3]   
4  [2, 5, 5, 5, 15, 5, 8, 6, 15, 15, 15, 15, 15, ...   

                                        dependencies  \
0  [[det, 2, 1], [nsubj, 5, 2], [cop, 5, 3], [amo...   
1  [[nsubj, 2, 1], [root, 0, 2], [aux, 4, 3], [xc...   
2  [[nsubj, 4, 1], [cop, 4, 2], [advmod, 4, 3], [...   
3  [[aux, 3, 1], [nsubj, 3, 2], [root, 0, 3], [ma...   
4  [[dep, 2, 1], [dep, 5, 2], [det, 5, 3], [amod,...   

               aspect_sentiment     from_to  
0           [[bread, positive]]    [[1, 2]]  
1  [[delivery times, positive]]  [[10, 12]]  
2            [[Food, positive]]    [[0, 1]]  
3          [[coffee, positive]]    [[5, 6]]  
4           [[place, positive]]  [[18, 19]]

In [12]:
test_depparse.columns

Index(['sentence', 'tokens', 'tags', 'dependencies', 'aspect_sentiment',
       'from_to'],
      dtype='object')

In [14]:
test_biaffine.columns

Index(['sentence', 'tokens', 'tags', 'predicted_dependencies',
       'predicted_heads', 'dependencies', 'aspect_sentiment', 'from_to'],
      dtype='object')

In [19]:
test_biaffine_energy.columns

Index(['sentence', 'tokens', 'tags', 'predicted_dependencies',
       'predicted_heads', 'dependencies', 'aspect_sentiment', 'from_to'],
      dtype='object')

In [49]:
test_biaffine_energy.iloc[0]['aspect_sentiment']


[['bread', 'positive']]

In [46]:
test_biaffine_energy.iloc[597]['aspect_sentiment']

[['food', 'negative'], ['waiter', 'negative']]

# Aspect Number

Aspect No = 1, 285 -- test_biaffine_energy_1

Aspect No = 2, 192 -- test_biaffine_energy_2

Aspect No >=3, 123 -- test_biaffine_energy_3


## Restaurants_Test_Gold_biaffine_depparsed_with_energy

In [97]:
test_biaffine_energy['aspect_no'] = test_biaffine_energy.apply(lambda row: len(row.aspect_sentiment), axis = 1)
test_biaffine_energy.groupby('aspect_no').count()

sentence  tokens  tags  predicted_dependencies  predicted_heads  \
aspect_no                                                                    
1               285     285   285                     285              285   
2               192     192   192                     192              192   
3                73      73    73                      73               73   
4                31      31    31                      31               31   
5                14      14    14                      14               14   
6                 3       3     3                       3                3   
7                 1       1     1                       1                1   
13                1       1     1                       1                1   

           dependencies  aspect_sentiment  from_to  sentiment_diff  \
aspect_no                                                            
1                   285               285      285             285   
2                   192               192      192             192   
3                    73                73       73              73   
4                    31                31       31              31   
5                    14                14       14              14   
6                     3                 3        3               3   
7                     1                 1        1               1   
13                    1                 1        1               1   

           best_sentiment  
aspect_no                  
1                     285  
2                     192  
3                      73  
4                      31  
5                      14  
6                       3  
7                       1  
13                      1

In [8]:
test_biaffine_energy_1 = test_biaffine_energy[['sentence', 'tokens', 'tags', 'predicted_dependencies',
       'predicted_heads', 'dependencies', 'aspect_sentiment', 'from_to']][test_biaffine_energy['aspect_no']==1]
test_biaffine_energy_1.to_json('Restaurants_Test_Gold_biaffine_depparsed_with_energy_1.json', orient='records')

In [9]:
test_biaffine_energy_2 = test_biaffine_energy[['sentence', 'tokens', 'tags', 'predicted_dependencies',
       'predicted_heads', 'dependencies', 'aspect_sentiment', 'from_to']][test_biaffine_energy['aspect_no']==2]
test_biaffine_energy_2.to_json('Restaurants_Test_Gold_biaffine_depparsed_with_energy_2.json', orient='records')

In [10]:
test_biaffine_energy_3 = test_biaffine_energy[['sentence', 'tokens', 'tags', 'predicted_dependencies',
       'predicted_heads', 'dependencies', 'aspect_sentiment', 'from_to']][test_biaffine_energy['aspect_no']>=3]
test_biaffine_energy_3.to_json('Restaurants_Test_Gold_biaffine_depparsed_with_energy_3.json', orient='records')

## Restaurants_Test_Gold_biaffine_depparsed

In [12]:
test_biaffine['aspect_no'] = test_biaffine.apply(lambda row: len(row.aspect_sentiment), axis = 1)
test_biaffine.groupby('aspect_no').count()

sentence  tokens  tags  predicted_dependencies  predicted_heads  \
aspect_no                                                                    
1               285     285   285                     285              285   
2               192     192   192                     192              192   
3                73      73    73                      73               73   
4                31      31    31                      31               31   
5                14      14    14                      14               14   
6                 3       3     3                       3                3   
7                 1       1     1                       1                1   
13                1       1     1                       1                1   

           dependencies  aspect_sentiment  from_to  
aspect_no                                           
1                   285               285      285  
2                   192               192      192  
3                    73                73       73  
4                    31                31       31  
5                    14                14       14  
6                     3                 3        3  
7                     1                 1        1  
13                    1                 1        1

In [13]:
test_biaffine1 = test_biaffine[['sentence', 'tokens', 'tags', 'predicted_dependencies',
       'predicted_heads', 'dependencies', 'aspect_sentiment', 'from_to']][test_biaffine['aspect_no']==1]
test_biaffine1.to_json('Restaurants_Test_Gold_biaffine_depparsed_1.json', orient='records')

In [14]:
test_biaffine2 = test_biaffine[['sentence', 'tokens', 'tags', 'predicted_dependencies',
       'predicted_heads', 'dependencies', 'aspect_sentiment', 'from_to']][test_biaffine['aspect_no']==2]
test_biaffine2.to_json('Restaurants_Test_Gold_biaffine_depparsed_2.json', orient='records')

In [15]:
test_biaffine3 = test_biaffine[['sentence', 'tokens', 'tags', 'predicted_dependencies',
       'predicted_heads', 'dependencies', 'aspect_sentiment', 'from_to']][test_biaffine['aspect_no']>=3]
test_biaffine3.to_json('Restaurants_Test_Gold_biaffine_depparsed_3.json', orient='records')

## Restaurants_Test_Gold_biaffine_depparsed

In [16]:
test_depparse['aspect_no'] = test_depparse.apply(lambda row: len(row.aspect_sentiment), axis = 1)
test_depparse.groupby('aspect_no').count()

sentence  tokens  tags  dependencies  aspect_sentiment  from_to
aspect_no                                                                 
1               285     285   285           285               285      285
2               192     192   192           192               192      192
3                73      73    73            73                73       73
4                31      31    31            31                31       31
5                14      14    14            14                14       14
6                 3       3     3             3                 3        3
7                 1       1     1             1                 1        1
13                1       1     1             1                 1        1

In [17]:
test_depparse1 = test_biaffine[['sentence', 'tokens', 'tags', 'predicted_dependencies',
       'predicted_heads', 'dependencies', 'aspect_sentiment', 'from_to']][test_depparse['aspect_no']==1]
test_depparse1.to_json('Restaurants_Test_Gold_depparsed_1.json', orient='records')

In [18]:
test_depparse2 = test_biaffine[['sentence', 'tokens', 'tags', 'predicted_dependencies',
       'predicted_heads', 'dependencies', 'aspect_sentiment', 'from_to']][test_depparse['aspect_no']==2]
test_depparse2.to_json('Restaurants_Test_Gold_depparsed_2.json', orient='records')

In [19]:
test_depparse3 = test_biaffine[['sentence', 'tokens', 'tags', 'predicted_dependencies',
       'predicted_heads', 'dependencies', 'aspect_sentiment', 'from_to']][test_depparse['aspect_no']>=3]
test_depparse3.to_json('Restaurants_Test_Gold_depparsed_3.json', orient='records')

# Aspect Sentiment

In [7]:
test_biaffine_energy.iloc[597]['aspect_sentiment']

[['food', 'negative'], ['waiter', 'negative']]

In [38]:
test_biaffine_energy.iloc[0]['aspect_sentiment']

[['bread', 'positive']]

In [42]:
def sent_code(data_row):
    '''
    positive = 1, negative = -1, other 0
    '''
    
    sentiments = [x[1] for x in data_row.aspect_sentiment]
    sentiment_code = [ 1 if item == 'positive' else -1 if item == 'negative' else 0 for item in sentiments]
    
    return sentiment_code



In [96]:
test_biaffine_energy['sentiment_diff'] = test_biaffine.apply(lambda row: max(sent_code(row))-min(sent_code(row)), axis =1)
test_biaffine_energy['best_sentiment'] = test_biaffine.apply(lambda row: max(sent_code(row)), axis =1)
test_biaffine_energy.groupby('sentiment_diff').count()

sentence  tokens  tags  predicted_dependencies  \
sentiment_diff                                                   
0                    520     520   520                     520   
1                     62      62    62                      62   
2                     18      18    18                      18   

                predicted_heads  dependencies  aspect_sentiment  from_to  \
sentiment_diff                                                             
0                           520           520               520      520   
1                            62            62                62       62   
2                            18            18                18       18   

                aspect_no  best_sentiment  
sentiment_diff                             
0                     520             520  
1                      62              62  
2                      18              18

Sentiment Difference 0, 520

Sentiment Difference 1, 62

Sentiment Difference 2, 18

-------


Multiple Aspects, Same sentiment  235

 - Multiple Aspects, All positive 177
 - Multiple Aspects, All negative 34
 - Multiple Aspects, All neutural 24

Multiple Aspects, Different sentiments 80



In [115]:
test_biaffine_energy_same_sentiment = test_biaffine_energy[['sentence', 'tokens', 'tags', 'predicted_dependencies',
       'predicted_heads', 'dependencies', 'aspect_sentiment', 'from_to']][(test_biaffine_energy['aspect_no']>1) & (test_biaffine_energy['sentiment_diff']==0)]

test_biaffine_energy_same_sentiment.to_json('Restaurants_Test_Gold_depparsed_with_energy_same_sentiment.json', orient='records')

In [73]:
test_biaffine_energy_same_sentiment.shape

(235, 8)

In [116]:
test_biaffine_energy_diff_sentiment = test_biaffine_energy[['sentence', 'tokens', 'tags', 'predicted_dependencies',
       'predicted_heads', 'dependencies', 'aspect_sentiment', 'from_to']][(test_biaffine_energy['aspect_no']>1) & (test_biaffine_energy['sentiment_diff']>0)]

test_biaffine_energy_diff_sentiment.to_json('Restaurants_Test_Gold_depparsed_with_energy_diff_sentiment.json', orient='records')

In [117]:
test_biaffine_energy_diff_sentiment.shape

(80, 8)

In [118]:
test_biaffine_energy_all_pos = test_biaffine_energy[['sentence', 'tokens', 'tags', 'predicted_dependencies',
       'predicted_heads', 'dependencies', 'aspect_sentiment', 'from_to']][(test_biaffine_energy['aspect_no']>1) & (test_biaffine_energy['sentiment_diff']==0) & (test_biaffine_energy['best_sentiment']== 1)]

test_biaffine_energy_all_pos.to_json('Restaurants_Test_Gold_depparsed_with_energy_all_pos.json', orient='records')

In [119]:
test_biaffine_energy_all_pos.shape

(177, 8)

In [120]:
test_biaffine_energy_all_neg = test_biaffine_energy[['sentence', 'tokens', 'tags', 'predicted_dependencies',
       'predicted_heads', 'dependencies', 'aspect_sentiment', 'from_to']][(test_biaffine_energy['aspect_no']>1) & (test_biaffine_energy['sentiment_diff']==0) & (test_biaffine_energy['best_sentiment']== -1)]
test_biaffine_energy_all_neg.to_json('Restaurants_Test_Gold_depparsed_with_energy_all_neg.json', orient='records')

In [121]:
test_biaffine_energy_all_neg.shape

(34, 8)

In [122]:
test_biaffine_energy_all_neu = test_biaffine_energy[['sentence', 'tokens', 'tags', 'predicted_dependencies',
       'predicted_heads', 'dependencies', 'aspect_sentiment', 'from_to']][(test_biaffine_energy['aspect_no']>1) & (test_biaffine_energy['sentiment_diff']==0) & (test_biaffine_energy['best_sentiment']==0)]

test_biaffine_energy_all_neu.to_json('Restaurants_Test_Gold_depparsed_with_energy_all_neu.json', orient='records')

In [123]:
test_biaffine_energy_all_neu.shape

(24, 8)